tests/functional/base.py

```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env python
+import os
+import shutil
+import subprocess
+import sys
+import time
+import unittest
+from string import Template
+
+import requests
+
+from custodia.client import HTTPUnixAdapter
+from custodia.server.args import parse_args
+from custodia.server.config import parse_config
+
+
+class CustodiaServerBase(unittest.TestCase):
+
+    request_headers = {'REMOTE_USER': 'me'}
+    test_dir = 'tests/functional/tmp'
+
+    @classmethod
+    def setUpClass(cls):
+        super(CustodiaServerBase, cls).setUpClass()
+
+    @classmethod
+    def create_server(cls, configfile=None, configstr=None):
+        if configfile is not None and configstr is not None:
+            raise ValueError
+        if configfile is not None:
+            with open(configfile) as f:
+                configstr = f.read()
+
+        if os.path.isdir(cls.test_dir):
+            shutil.rmtree(cls.test_dir)
+        os.makedirs(cls.test_dir)
+        cls.configstr = configstr
+        custodia_conf = os.path.join(cls.test_dir, 'custodia.conf')
+        with (open(custodia_conf, 'w+')) as conffile:
+            t = Template(cls.configstr)
+            conf = t.substitute({'TEST_DIR': cls.test_dir})
+            conffile.write(conf)
+        cls.args = parse_args([custodia_conf])
+        cls.parser, cls.config = parse_config(cls.args)
+        cls.session = requests.Session()
+        cls.session.mount('http+unix://', HTTPUnixAdapter())
+        env = os.environ.copy()
+        env['CUSTODIA_TMPDIR'] = cls.test_dir
+        env['CUSTODIA_SOCKET'] = cls.config['socketdir']
+        cls.process = subprocess.Popen(
+            [sys.executable, '-m', 'custodia.server', custodia_conf]
+        )
+        time.sleep(1)
```

Comment:
Magic sleeps are problematic because they lead to flaky tests. Please use a loop that looks for failing process and waits for the server socket to be available:


Pseudo code:
```
timeout = time.time() + 5
while time.time() < timeout:
    if cls.process.poll() is None:
        raise AssertionError(...)
    try:
         with socket.socket(family=socket.AF_UNIX) as s:
              s.connect(env['CUSTODIA_SOCKET'])
     except OSError:
         pass
      else:
           return True
    time.sleep(0.1)
raise Error("timeout")
```